sympy/polys/matrices/dense.py

```diff
@@ -116,38 +116,33 @@ def ddm_irref(a, _partial_pivot=False):
         # domain is RR or CC. It uses partial (row) pivoting based on the
         # absolute value of the pivot candidates.
         if _partial_pivot:
-            ip = max(range(i, m), key=lambda ip: abs(a[ip][j]))
-            a[i], a[ip] = a[ip], a[i]
-
-        # pivot
-        aij = a[i][j]
+            p = max(range(i, m), key=lambda ip: abs(a[ip][j]))
+            a[i], a[p] = a[p], a[i]
 
         # zero-pivot
-        if not aij:
-            for ip in range(i+1, m):
-                aij = a[ip][j]
+        if not a[i][j]:
+            for p in range(i+1, m):
                 # row-swap
-                if aij:
-                    a[i], a[ip] = a[ip], a[i]
+                if a[p][j]:
+                    a[i], a[p] = a[p], a[i]
                     break
             else:
                 # next column
                 continue
 
         # normalise row
         ai = a[i]
-        aijinv = aij**-1
-        for l in range(j, n):
-            ai[l] *= aijinv # ai[j] = one
+        for l in range(j + 1, n):
+            ai[l] /= ai[j]
+        ai[j] /= ai[j]
```

Comment:
Be very careful changing this and make sure to run timings with domains like `QQ(x, y)` or `QQ(sqrt(2))` etc. I'm not happy with exactly how this code works (#21834) but don't even think about changing this for reasons of aesthetics or micro-optimisation without a detailed investigation.

Reply:
Master:
```python
In [1]: from sympy import *
   ...: from sympy.polys.matrices import *
   ...: 
   ...: K = QQ.algebraic_field(sqrt(2))
   ...: for n in range(2, 10):
   ...:     M = Matrix(n, n, lambda i, j: (i**2+j+j*i)**10*sqrt(2) + j**2)
   ...:     dM = M._rep.convert_to(K)
   ...:     %time dM.inv()
   ...: 
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 787 µs
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 1.82 ms
CPU times: user 4 ms, sys: 0 ns, total: 4 ms
Wall time: 4.17 ms
CPU times: user 8 ms, sys: 0 ns, total: 8 ms
Wall time: 8.37 ms
CPU times: user 16 ms, sys: 0 ns, total: 16 ms
Wall time: 15.2 ms
CPU times: user 28 ms, sys: 0 ns, total: 28 ms
Wall time: 25.5 ms
CPU times: user 40 ms, sys: 0 ns, total: 40 ms
Wall time: 39.5 ms
CPU times: user 60 ms, sys: 0 ns, total: 60 ms
Wall time: 59.1 ms
```
PR:
```python
In [1]: from sympy import *
   ...: from sympy.polys.matrices import *
   ...: 
   ...: K = QQ.algebraic_field(sqrt(2))
   ...: for n in range(2, 10):
   ...:     M = Matrix(n, n, lambda i, j: (i**2+j+j*i)**10*sqrt(2) + j**2)
   ...:     dM = M._rep.convert_to(K)
   ...:     %time dM.inv()
   ...: 
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 1.36 ms
CPU times: user 4 ms, sys: 0 ns, total: 4 ms
Wall time: 3.29 ms
CPU times: user 8 ms, sys: 0 ns, total: 8 ms
Wall time: 6.86 ms
CPU times: user 12 ms, sys: 0 ns, total: 12 ms
Wall time: 12.7 ms
CPU times: user 20 ms, sys: 0 ns, total: 20 ms
Wall time: 22 ms
CPU times: user 32 ms, sys: 0 ns, total: 32 ms
Wall time: 34.6 ms
CPU times: user 52 ms, sys: 0 ns, total: 52 ms
Wall time: 51.9 ms
CPU times: user 76 ms, sys: 0 ns, total: 76 ms
Wall time: 74.8 ms
```
Timings with CPython 3.8 on Linux.